### ISSUE TITLE
Slow processing of large multi instance collection

### ISSUE THREAD

**Issue by ChrisKujawa** [camunda/camunda#12946]
**Describe the bug**

As discovered in the [recent chaos day](https://zeebe-io.github.io/zeebe-chaos/2023/06/02/Using-Large-Multi-Instance/) it seems that the current batching of large multi-instance slows the processing down by a lot. It worked in general which is great!, but it was quite slow.

It might happen that a large input collection could consume the complete processing of a partition.

In the current state, it is not clear where the issue lies, but it seems that the batch processing is limited to just 2-4 commands per batch, which slows down the creation of instances.

We should further investigate whether this issue is related to batch processing itself, or based on new instance batching.

**To Reproduce**

Use https://github.com/camunda-cloud/game-day-cookbook/tree/main/mains/large-multi-instance in order to reproduce this

**Expected behavior**

I somehow would expect that the instance creation is faster and makes more use of the batch processing.

For further details take a look at the chaos day summary.

**Comment by megglos** [comment]
ZDP-Triage:
- it's not clear yet at what number of multi-instances the performance degradation becomes noticeable
- degrades performance of the whole partition (took 10m for creating 20k instances)
- cancellation might cause the same performance hit
- there is no workaround, it eventually completes but requires a considerable amount of time
- might be caused by a bug in batch processing => we should validate this first to decide on how to move on

**Comment by remcowesterhoud** [comment]
> link the recent fix introduced that surfaced this behavior

It is #12692 :)

**Comment by korthout** [comment]
ZPA triage:
- a lower batch size may improve the performance similar to message ttl checking / due date timer checking
- we want to have a look at limiting the batch size with a configurable number 
- at least it does work now, but degraded performance is not great
- as a workaround the previous workaround of modeling batching in the process is still a possibility to avoid the degraded performance
- priority is `later`

**Comment by megglos** [comment]
ZDP-Planning:
- it results in a bad user experience as internal commands cause a high load leading to backpressure of user requests even if user requests are at a very low rate (affected a customer already)
- workaround available being lowering batch size only has a minimal effect
- `might be caused by a bug in batch processing => we should validate this first to decide on how to move on` => @Zelldon let's validate this next

**Comment by Zelldon** [comment]
# Investigation :detective: 

I had a deeper look at this issue. I think I know understand why this happens.

**TL;DR;** It is indeed related to the batch processing, but also the combination with multi-instances and the new batching approach in the engine.

## Details

I have played a bit with a unit test we have an created a large input collection: ` IntStream.range(1, 100_000).boxed().toList();`

_Example process:_

```
Bpmn.createExecutableProcess(PROCESS_ID)
        .startEvent()
        .serviceTask(
            ELEMENT_ID,
            t -> t.zeebeJobType(jobType).multiInstance(INPUT_VARIABLE_BUILDER.andThen(builder)))
        .endEvent()
        .done();
```


### Scenario

**ServiceTask with a multi-instance marker**

[In the case of a parallel multi-instance we write a ProcessInstanceBatchRecord](https://github.com/camunda/zeebe/blob/main/engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/container/MultiInstanceBodyProcessor.java#L277). This will be processed as part of the batch.

The corresponding [ActivateProcessInstanceBatchProcessor](https://github.com/camunda/zeebe/blob/main/engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ActivateProcessInstanceBatchProcessor.java#L48) will add ACTIVATE commands to the batch, until the batch is full. Here the first issue appears.

Let's say we have the following batch now: `[..., PI Batch Activate (index: 100000), Activate Element, Activate Element, Activate Element, ..., PI Batch Activate (index: 89021)]` Due to the logic in the `ActivateProcessInstanceBatchProcessor` we write the batch almost full, at the end we add another ProcessInstanceBatchRecord with the remaining elements.

In the ProcessingStateMachine, where batch processing happens, we will try to process the next command (after the first ProcessInstanceBatchRecord). This is the first activate element command (for the multi-instance service task), it might succeed for the first elements. At some point, we will fail with ExceedBatchSize, rollback, and [try until the position where it last time failed to write at least that batch](https://github.com/camunda/zeebe/blob/main/stream-platform/src/main/java/io/camunda/zeebe/stream/impl/ProcessingStateMachine.java#L325-L331). 

_This means we will always have that retry mechanism triggered._ 

**The most problematic part comes next.**

After we have completed that previous batch, we will read one command at a time, which is in this case the ACTIVATE element command. It is likely a wait state (potentially a service task) which means there will be no more follow-up commands (except the Job create). We are doing this for all commands which are then written in between these ProcessInstanceBatchRecords we have seen above.

This means the problem here is that we are not batching persisted commands, and processing them one-by-one which slows the process instance execution enormously down. This explains also why the partition seems to be blocked and no other instance might make progress, since such big batches are written to the log. 



### Potential solutions

1. Batch processing can be more clever and detect already earlier that the batch is already full enough. To avoid the retry.
2. Reading batch of commands, when they are for the same process instance. 
3. Command batches can be more clever, instead writing of 1000+x commands just write one, which says to activate 1000 elements. The event can contain all keys but might defer the problem since we still need to write the next command like create a job. But maybe this can be combined as well. Would be interesting to investigate.

**Comment by megglos** [comment]
ZDP-Planning:
- effectively batch processing has no effect on the activation commands
- moving to backlog as upcoming to improve performance for such scenarios

**Comment by mustafadagher** [comment]
We have experienced a similar issue again today in a `Game Day`, and I suggest until solving this issue fundamentally, we should do two things.
1. run a benchmark test to find out at what limit does our performance start do deteriorate, and update our docs to have guidelines to stay below this limit.
2. create an incident if a user tries to exceed this limit until we unblock this.

**Comment by deepthidevaki** [comment]
Here is a [support ticket](https://jira.camunda.com/browse/SUPPORT-22755) affected by this issue. 

Moving this issue from backlog so that the team can discuss this.

**Comment by npepinpe** [comment]
Felix is on board about a proper fix, and we're currently checking with the customer about a timeline. I've proposed after 8.7 at the moment, so May 2025. That does mean another 8 months with the current status quo. Is this something we can live with, or do we need a band-aid in the meantime? We already sort of have one for users wishing to cancel via force banning, though it's not perfect.

Do we think we would find capacity to actually implement the workaround proposed in #19531? Because my gut tells me - likely not. As for what Chris attempted, I'm not sure I understand it, so I can't say whether it's worth exploring more (since his last status was that it was more complex than expected), how much it would help, and if it helps reach a proper solution at all (or if it's work that will be thrown out in the end).

**Comment by jothikiruthika** [comment]
Here is another [support ticket ](https://jira.camunda.com/browse/SUPPORT-22888)related to the same concept - ~6k multi-instance loops in a single PI. Customer observed that they triggered only 5-7 PI and their Dev cluster became too slow.

**Comment by korthout** [comment]
In September, both Zeebe teams met in person in the Berlin office for a workshop/team event. The team worked on multiple problems and this was one of them. I wanted to share the findings here for posterity.

[This Google Docs page](https://docs.google.com/document/d/1D4nJS7zsYUBe5s0sodzRR2N57eLu0HQBn1eqQ2-roow/edit?usp=sharing) contains the notes of this workshop. It is publicly accessible as it doesn't share anything that we wouldn't have written here on GitHub. However, note that the notes are quite raw and have not been edited for readability. It's just available to help guide the solution design of this problem.

**Comment by korthout** [comment]
Added a target because of [this message](https://jira.camunda.com/browse/SUPPORT-22755?focusedCommentId=362623&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-362623)

**Comment by korthout** [comment]
Today, @remcowesterhoud @eppdot @koevskinikola @tmetzke @ana-vinogradova-camunda and I looked into this issue during our mob programming:
- we explored the basic idea [presented previously](https://github.com/camunda/camunda/issues/12946#issuecomment-2478463119)
- we explored an alternative idea to not write the `ProcessInstanceBatchRecord` at all, by letting the `MultiInstanceBodyProcessor` take care of writing a `ProcessInstanceRecord` for the next child instance to activate
  - it could do so `onChildActivating`
  - however, we need to know the total input collection size which requires expression evaluation
  - this can fail and produce an incident, in which case we cannot activate the next child instance
  - in other words, the multi-instance activation is stuck if there's an incident there
  - this could be solved if we somehow store the input collection/size, but we currently don't
- we decided that next time, we'll continue with the simpler solution: just reduce the `ProcessInstanceBatch` limit to 1 as a quick fix
- we could investigate further optimizations in the future when that need arises
